### ISSUE TITLE
Race condition at shutdown in ddtrace 2.11.1 (possibly later.)

### ISSUE THREAD

**Issue by antarus12345** [DataDog/dd-trace-py#11622]
TL;DR, a subset of my programs failed to exit once program termination is triggered. I believe there exists a race condition in ddtrace (2.11.1 and it looks like in master) on program termination. The race seems fairly rare, but my programs somehow hit it consistently enough.

The bug results from basically 3 threads executing.

1. The main program (T0)
2. A background `periodicThread` that is datadog.telemetry.writer that is running in the background sending telemetry to datadog (T1)
3. Some other thread that is throwing an exception near program termination and is running `ddtrace.telemetry.writer.telemetry_excepthook` (T2)

The basic idea is that once the program ends, python runs the atexit handlers in LIFO order. It calls telemetrywriter.atexit() which is supposed to take the telemetry writer lock and then stop the telemetry writer. However, some thread T2 is in the middle of telemetrywriter.excepthook() and is holding the lock (inside of app_shutdown) This causes T0 to block on acquiring the lock that T1 is holding.

However T1 is calling app_shutdown, which is trying to stop the TelemtryWriter service. This needs T2 to exit. Unfortunately, T2 will run until it is stopped by its own atexit handler.

T0 main thread running telememtryWwiter.atexit(), blocking on telemetryWriterservice_lock (held by t1)
T1 worker thread running telemetrywriter.excepthook(), holding telemetryWriterservice_lock, waiting on T2 to complete
T2 background thread running telemetrywriter.periodicThread() will run forever until its atexit is called, but will not be called until T1 finishes its current atexit handler, which will never happen because T0 and T1 are in deadlock.

So to trigger we need I think at least:
 - 2 program threads.
 - Exception to be happening near program exit.
 - the Periodic Thread's atexit handler has to be registered *before* the telemetryWriter's atexit (because atexits are run in LIFO order.)

I'm still working on a clean reproducer for this :/

**Comment by antarus12345** [comment]
(as an obvious A/B test for me I will attempt to set DD_INSTRUMENTATION_TELEMETRY_ENABLED to false in my  staging environment to see if the problem persists; this would in theory prevent the app_shutdown atexit from being registered and would prevent the deadlock.

**Comment by mabdinur** [comment]
Hey @antarus12345, 

Thanks for bringing this issue to our attention. I wasn't able to reproduce this issue but we can prevent this deadlock by not stopping the periodic thread in [app_shutdown]((https://github.com/DataDog/dd-trace-py/blob/v2.17.3/ddtrace/internal/telemetry/writer.py#L644). The ddtrace library already registers an atexit hook that halts all periodic threads [here](https://github.com/DataDog/dd-trace-py/blob/v2.17.3/ddtrace/internal/periodic.py#L19), we do not need to duplicate this logic in app_shutdown.


We can also remove all usages of TelemetryWriter._lock ([ex](https://github.com/DataDog/dd-trace-py/blob/v2.17.3/ddtrace/internal/telemetry/writer.py#L644)) and instead use the [service_lock](https://github.com/DataDog/dd-trace-py/blob/v2.17.3/ddtrace/internal/service.py#L82) for all critical operations (starting/stopping/awaking threads, queuing events, etc.). This should reduce the likelihood of deadlocks.
